### PR TITLE
7881 explicit persist side effects layer+widgets save

### PR DESCRIPTION
**Pull request by viddo**
Milestone for #8042

Added user-actions to save layer and widget (and differentiate it from saving widget-option.

Also search-and-replace’d some methods that had a bit unorthodox names.

@alonsogarciapablo review, please :bow:

### CODE REVIEW

File: lib/assets/javascripts/cartodb3/editor/widgets/widgets-view.js [hunk 1]
```diff
@@ -60,7 +60,7 @@ module.exports = CoreView.extend({
     this.$('.js-widgets > .js-widgetItem').each(function (index, item) {
       var modelCid = $(item).data('model-cid');
       var model = self._widgetDefinitionsCollection.get(modelCid);
-      model.save({ order: index });
+      model.save({ order: index }); // TODO userAction.saveWidget
```
Comment (alonsogarciapablo) [inline lib/assets/javascripts/cartodb3/editor/widgets/widgets-view.js]: Left a TODO here...

Comment (viddo) [inline lib/assets/javascripts/cartodb3/editor/widgets/widgets-view.js]: That was totally on purpose to see if you were awake 🙈

File: lib/assets/javascripts/cartodb3/data/user-actions.js [hunk 2]
```diff
@@ -104,14 +104,30 @@ module.exports = function (collections) {
...
+    },
+
+    saveWidget: function (widgetDefModel) {
+      if (!widgetDefModel) throw new Error('widgetDefModel is required');
+
+      var widgetLayerId = widgetDefModel.get('layer_id');
+      layerDefinitionsCollection.some(function (layerDefModel) {
+        if (layerDefModel.id === widgetLayerId) {
+          this.saveLayer(layerDefModel);
+          return true; // aborts the "some"-iterator
```
Comment (alonsogarciapablo) [inline lib/assets/javascripts/cartodb3/data/user-actions.js]: Is this why we're using `.some` instead of `.each`? (just curious)

Comment (viddo) [inline lib/assets/javascripts/cartodb3/data/user-actions.js]: Yes, to stop iterating once found the associated layer

File: lib/assets/javascripts/cartodb3/data/analysis-definitions-collection.js [hunk 3]
```diff
@@ -60,6 +60,7 @@ module.exports = Backbone.Collection.extend({
    * @return {jqXHR, undefined} A promise if created a new analysis
    */
   saveAnalysis: function (nodeDefModel, opts) {
+    if (!nodeDefModel) throw new Error('nodeDefModel is required');
```
Comment (alonsogarciapablo) [inline lib/assets/javascripts/cartodb3/data/analysis-definitions-collection.js]: Are we planning  this logic (👇) to user-actions.js too? (it seems to have some side-effects too)

**Comment by alonsogarciapablo** [comment]
Looks good to me! 👍

File: lib/assets/javascripts/cartodb3/data/analysis-definitions-collection.js (hunk 3 above)
Comment (viddo) [inline lib/assets/javascripts/cartodb3/data/analysis-definitions-collection.js]: It's already being used there, but probably makes sense to move it over there, I'll do that separately though.

**Pull request merged**